src/main/scheduler/scheduler.c

```diff
@@ -34,36 +36,60 @@
 #include "common/utils.h"
 
 #include "drivers/time.h"
+#include "drivers/accgyro/accgyro.h"
+#include "drivers/system.h"
 
 #include "fc/core.h"
 #include "fc/tasks.h"
 
 #include "scheduler.h"
 
-#define TASK_AVERAGE_EXECUTE_FALLBACK_US 30 // Default task average time if USE_TASK_STATISTICS is not defined
-#define TASK_AVERAGE_EXECUTE_PADDING_US 5   // Add a little padding to the average execution time
+#include "sensors/gyro_init.h"
 
 // DEBUG_SCHEDULER, timings for:
 // 0 - gyroUpdate()
 // 1 - pidController()
 // 2 - time spent in scheduler
 // 3 - time spent executing check function
 
+// DEBUG_DETERMINISM, requires USE_LATE_TASK_STATISTICS to be defined
+// 0 - Gyro task start cycle time in 10th of a us
+// 1 - ID of late task
+// 2 - Amount task is late in 10th of a us
+// 3 - Gyro lock skew in clock cycles
+
+extern task_t tasks[];
+
 static FAST_DATA_ZERO_INIT task_t *currentTask = NULL;
-static FAST_DATA_ZERO_INIT bool ignoreCurrentTaskTime;
+static FAST_DATA_ZERO_INIT bool ignoreCurrentTaskExecRate;
+static FAST_DATA_ZERO_INIT bool ignoreCurrentTaskExecTime;
+
+int32_t schedLoopStartCycles;
+static int32_t schedLoopStartMinCycles;
+static int32_t schedLoopStartMaxCycles;
+static int32_t schedLoopStartDeltaDownCycles;
+static int32_t schedLoopStartDeltaUpCycles;
+
+int32_t taskGuardCycles;
+static int32_t taskGuardMinCycles;
+static int32_t taskGuardMaxCycles;
+static int32_t taskGuardDeltaDownCycles;
+static int32_t taskGuardDeltaUpCycles;
 
 static FAST_DATA_ZERO_INIT uint32_t totalWaitingTasks;
 static FAST_DATA_ZERO_INIT uint32_t totalWaitingTasksSamples;
 
-static FAST_DATA_ZERO_INIT bool calculateTaskStatistics;
 FAST_DATA_ZERO_INIT uint16_t averageSystemLoadPercent = 0;
 
 static FAST_DATA_ZERO_INIT int taskQueuePos = 0;
 STATIC_UNIT_TESTED FAST_DATA_ZERO_INIT int taskQueueSize = 0;
 
-static FAST_DATA int periodCalculationBasisOffset = offsetof(task_t, lastExecutedAtUs);
+static bool optimizeSchedRate;
```

Comment:
What does 'Sched' stand for?

Reply:
Scheduler. Trying to keep the variable name lengths sane. Happy to expand if you'd prefer.